File: Archive_LJ_grand_canonical/gcmclj.f90
!**********************************************************************!
!
! File: gcmclj.f90
!
! Grand Canonical (TVmu) Monte Carlo of Lennard-Jonesium
!
! 04-Feb-2004 (MN)
! 19-Apr-2012
!
!**********************************************************************!

module gcmclj_glbm

!**********************************************************************!

! Parameters & global variables

  implicit none

  private

  integer,parameter,public::long=selected_int_kind(18)
  integer,parameter,public::double=selected_real_kind(15)
  integer,parameter,public::iin=1,iout=2

  integer,public::n,nt,ntjob,ntprint,ntskip
  integer,public::naccc,naccd,naccm,ndr,nmax
  integer,dimension(:),allocatable,public::iran
  real,public::disp,dr,pcre,t,v,zz
  real,public::alnmax,c,c2,c2m1,drm1,pi,r2max,rc2,su0,sw0,zv
  real,dimension(:),allocatable,public::x,y,z
  integer(kind=long),dimension(:),allocatable,public::ag
  real(kind=double),public::accrc,accrd,accrm,an,an2,au,aw

end module gcmclj_glbm

!**********************************************************************!

module gcmclj_subm

!**********************************************************************!

  use gcmclj_glbm

  implicit none

  private

  public::create,delete,getcf,means,move,putcf

contains

!**********************************************************************!

  subroutine getcf()

!**********************************************************************!

    integer::nran
    real::rc

! Maximum argument for exponential & pi

    alnmax=log(0.1*huge(1.0))
    pi=4.0*atan(1.0)

! RNG characteristics

    call random_seed(size=nran)
    allocate(iran(nran))

! Read startup/checkpoint file

    open(unit=iin,file="gcmclj_in.dat",status="old",action="read", &
         form="unformatted",position="rewind")

    read(unit=iin) n,nt,ntjob,ntprint,ntskip,iran
    read(unit=iin) disp,dr,pcre,t,v,zz

! Allocate arrays

    ndr=int(0.5*v**(1.0/3.0)/dr)

    allocate(ag(0:ndr-1),x(n),y(n),z(n))

! Positions

    read(unit=iin) x(1:n),y(1:n),z(1:n)

! Read accumulated averages/clear accumulators

    if(nt>0) then
      read(unit=iin) accrc,accrd,accrm,an,an2,au,aw
      read(unit=iin) ag(0:ndr-1)
    else
      accrc=0.0_double
      accrd=0.0_double
      accrm=0.0_double
      an=0.0_double
      an2=0.0_double
      au=0.0_double
      aw=0.0_double
      ag=0_long
    end if

    close(unit=iin)

! Parameters

    nmax=n

    c=v**(1.0/3.0)
    c2=0.5*c
    c2m1=2.0/c
    drm1=1.0/dr
    r2max=c2**2
    zv=zz*v

! Cutoff & tail corrections

    rc=0.5*c
    rc2=rc**2

    su0=2.0*pi/v*(4.0/(9.0*rc**9)-4.0/(3.0*rc**3))
    sw0=2.0*pi/v*(24.0/(3.0*rc**3)-48.0/(9.0*rc**9))

! Initialize RNG

    call random_seed(put=iran)

! Clear acceptance counters

    naccc=0
    naccd=0
    naccm=0

    return

  end subroutine getcf

!**********************************************************************!

  subroutine create()

!**********************************************************************!

    logical::accept
    integer::j
    real::dx,dy,dz,r2,rm6,ran,su,xn,yn,zn
    real,dimension(:),allocatable::xt,yt,zt

! Random position

    call random_number(ran)
    xn=(ran-0.5)*c
    call random_number(ran)
    yn=(ran-0.5)*c
    call random_number(ran)
    zn=(ran-0.5)*c

! Energy difference

    su=(2*n+1)*su0

    do j=1,n
      dx=x(j)-xn
      dx=dx-int(dx*c2m1)*c
      dy=y(j)-yn
      dy=dy-int(dy*c2m1)*c
      dz=z(j)-zn
      dz=dz-int(dz*c2m1)*c
      r2=dx**2+dy**2+dz**2
      if(r2<rc2) then
        rm6=1.0/r2**3
        su=su+(4.0*rm6-4.0)*rm6
      end if
    end do

! Acceptance test

    if(su/t>alnmax) then
      accept=.false.
    else
      if(zv*exp(-su/t)/(n+1)>=1.0) then
        accept=.true.
      else
        call random_number(ran)
        if(ran<=zv*exp(-su/t)/(n+1)) then
          accept=.true.
        else
          accept=.false.
        end if
      end if
    end if

! Accept move & append particle to arrays

    if(accept) then
      naccc=naccc+1

! Extend arrays

      if(n>=nmax) then
        allocate(xt(n),yt(n),zt(n))
        xt=x(1:n)
        yt=y(1:n)
        zt=z(1:n)
        deallocate(x,y,z)
        nmax=n+1
        allocate(x(nmax),y(nmax),z(nmax))
        x(1:n)=xt
        y(1:n)=yt
        z(1:n)=zt
        deallocate(xt,yt,zt)
      end if

      n=n+1
      x(n)=xn
      y(n)=yn
      z(n)=zn

    end if

    return

  end subroutine create

!**********************************************************************!

  subroutine move()

!**********************************************************************!

    logical::accept
    integer::i,j
    real::dx,dy,dz,r2,rm6,ran,su,xn,yn,zn

! Select random particle

    call random_number(ran)
    i=min(int(n*ran+1.0),n)

! Trial move (displacement)

    call random_number(ran)
    xn=x(i)+(ran-0.5)*disp
    xn=xn-int(xn*c2m1)*c
    call random_number(ran)
    yn=y(i)+(ran-0.5)*disp
    yn=yn-int(yn*c2m1)*c
    call random_number(ran)
    zn=z(i)+(ran-0.5)*disp
    zn=zn-int(zn*c2m1)*c

! Energy difference

    su=0.0

    do j=1,n
      if(j/=i) then

! New position

        dx=x(j)-xn
        dx=dx-int(dx*c2m1)*c
        dy=y(j)-yn
        dy=dy-int(dy*c2m1)*c
        dz=z(j)-zn
        dz=dz-int(dz*c2m1)*c
        r2=dx**2+dy**2+dz**2
        if(r2<rc2) then
          rm6=1.0/r2**3
          su=su+(4.0*rm6-4.0)*rm6
        end if

! Old position

        dx=x(j)-x(i)
        dx=dx-int(dx*c2m1)*c
        dy=y(j)-y(i)
        dy=dy-int(dy*c2m1)*c
        dz=z(j)-z(i)
        dz=dz-int(dz*c2m1)*c
        r2=dx**2+dy**2+dz**2
        if(r2<rc2) then
          rm6=1.0/r2**3
          su=su-(4.0*rm6-4.0)*rm6
        end if

      end if
    end do

! Acceptance test

    if(su<=0.0) then
      accept=.true.
    else
      if(su/t>alnmax) then
        accept=.false.
      else
        call random_number(ran)
        if(ran<=exp(-su/t)) then
          accept=.true.
        else
          accept=.false.
        end if
      end if
    end if

! Accept move

    if(accept) then
      naccm=naccm+1
      x(i)=xn
      y(i)=yn
      z(i)=zn
    end if

    return

  end subroutine move

!**********************************************************************!

  subroutine delete()

!**********************************************************************!

    logical::accept
    integer::i,j
    real::dx,dy,dz,r2,rm6,ran,su

! Empty box?

    if(n<1) then
      return
    end if

! Random particle

    call random_number(ran)
    i=min(int(n*ran+1.0),n)

! Energy difference (ok for n=1)

    su=-(2*n-1)*su0

    do j=1,n
      if(j/=i) then
        dx=x(j)-x(i)
        dx=dx-int(dx*c2m1)*c
        dy=y(j)-y(i)
        dy=dy-int(dy*c2m1)*c
        dz=z(j)-z(i)
        dz=dz-int(dz*c2m1)*c
        r2=dx**2+dy**2+dz**2
        if(r2<rc2) then
          rm6=1.0/r2**3
          su=su-(4.0*rm6-4.0)*rm6
        end if
      end if
    end do

! Acceptance test

    if(su/t>alnmax) then
      accept=.false.
    else
      if(n*exp(-su/t)/zv>=1.0) then
        accept=.true.
      else
        call random_number(ran)
        if(ran<=n*exp(-su/t)/zv) then
          accept=.true.
        else
          accept=.false.
        end if
      end if
    end if

! Accept move & eliminate particle from arrays

    if(accept) then
      naccd=naccd+1
      x(i)=x(n)
      y(i)=y(n)
      z(i)=z(n)
      n=n-1
    end if

    return

  end subroutine delete

!**********************************************************************!

  subroutine means()

!**********************************************************************!

    integer::i,ir,j
    real::dx,dy,dz,r2,rm6,su,sw

! Potential energy, virial & g(r)

    su=n**2*su0
    sw=n**2*sw0

    do i=1,n-1
      do j=i+1,n

        dx=x(j)-x(i)
        dx=dx-int(dx*c2m1)*c
        dy=y(j)-y(i)
        dy=dy-int(dy*c2m1)*c
        dz=z(j)-z(i)
        dz=dz-int(dz*c2m1)*c
        r2=dx**2+dy**2+dz**2

        if(r2<rc2) then
          rm6=1.0/r2**3
          su=su+(4.0*rm6-4.0)*rm6
          sw=sw+(24.0-48.0*rm6)*rm6
        end if

        if(r2<r2max) then
          ir=int(drm1*sqrt(r2))
          if(ir<ndr) then
            ag(ir)=ag(ir)+1_long
          end if
        end if

      end do
    end do

! Print control variables

    if(ntprint>0.and.modulo(nt,ntprint)==0) then
      write(unit=*,fmt=*) nt,naccc/max(pcre*n*ntskip,1.0), &
           naccd/max(pcre*n*ntskip,1.0), &
           naccm/max((1.0-2.0*pcre)*n*ntskip,1.0), &
           n,n/v,su/n
    end if

! Accumulate averages

    accrc=accrc+naccc/max(pcre*n*ntskip,1.0)
    accrd=accrd+naccd/max(pcre*n*ntskip,1.0)
    accrm=accrm+naccm/max((1.0-2.0*pcre)*n*ntskip,1.0)
    an=an+n
    an2=an2+n**2
    au=au+su
    aw=aw+sw

! Clear acceptance counters

    naccc=0
    naccd=0
    naccm=0

    return

  end subroutine means

!**********************************************************************!

  subroutine putcf()

!**********************************************************************!

! RNG seed

    call random_seed(get=iran)

! Write checkpoint file

    open(unit=iout,file="gcmclj_out.dat",status="replace", &
         action="write",form="unformatted")

    write(unit=iout) n,nt,ntjob,ntprint,ntskip,iran
    write(unit=iout) disp,dr,pcre,t,v,zz
    write(unit=iout) x(1:n),y(1:n),z(1:n)
    write(unit=iout) accrc,accrd,accrm,an,an2,au,aw
    write(unit=iout) ag

    close(unit=iout)

! Deallocate arrays

    deallocate(ag,iran,x,y,z)

    return

  end subroutine putcf

end module gcmclj_subm

!**********************************************************************!

program gcmclj

!**********************************************************************!

  use gcmclj_glbm
  use gcmclj_subm

  implicit none

  integer::i,j
  real::ran

! Read startup/checkpoint file & initialize variables

  call getcf()

! Do (ntskip*ntjob) passes (particle creations/deletions/moves)

  do i=1,ntjob
    nt=nt+1
    do j=1,n*ntskip
      call random_number(ran)
      if(ran<pcre) then
        call create()            ! Create particle
      else if(ran<2.0*pcre) then
        call delete()            ! Delete particle
      else
        call move()              ! Move particle
      end if
    end do
    call means()
  end do

! Write checkpoint file

  call putcf()

end program gcmclj

!**********************************************************************!

File: Archive_LJ_grand_canonical/gcmclj.c
/**********************************************************************
 *
 * File: gcmclj.c
 *
 * Grand Canonical (TVmu) Monte Carlo of Lennard-Jonesium
 *
 * 09-Apr-2010 (MN)
 * 19-Apr-2012
 *
 **********************************************************************/

#include <stdio.h>
#include <stdlib.h>
#include <float.h>
#include <math.h>

#define TRUE 1
#define FALSE 0
#define max(a,b) ((a)>(b)?(a):(b))
#define min(a,b) ((a)<(b)?(a):(b))

/* Parameters & global variables */

int n,nt,ntjob,ntprint,ntskip;
int naccc,naccd,naccm,ndr,nmax;
double disp,dr,pcre,t,v,zz;
double alnmax,c,c2,c2m1,drm1,r2max,rc,rc2,su0,sw0,zv;
double *x,*y,*z;
long long *ag;
double accrc,accrd,accrm,an,an2,au,aw;

/**********************************************************************/

void getcf() {

/**********************************************************************/

  FILE *fpi;
  unsigned short seed[3];
  int i;

  /* Maximum argument for exponential */

  alnmax=log(0.1*DBL_MAX);

  /* Read startup/checkpoint file */

  fpi=fopen("gcmclj_in.dat","r");

  fread(&n,sizeof(int),1,fpi);
  fread(&nt,sizeof(int),1,fpi);
  fread(&ntjob,sizeof(int),1,fpi);
  fread(&ntprint,sizeof(int),1,fpi);
  fread(&ntskip,sizeof(int),1,fpi);
  fread(seed,sizeof(unsigned short),3,fpi);

  fread(&disp,sizeof(double),1,fpi);
  fread(&dr,sizeof(double),1,fpi);
  fread(&pcre,sizeof(double),1,fpi);
  fread(&t,sizeof(double),1,fpi);
  fread(&v,sizeof(double),1,fpi);
  fread(&zz,sizeof(double),1,fpi);

  /* Allocate arrays */

  ndr=0.5*cbrt(v)/dr;

  ag=(long long*)malloc(ndr*sizeof(long long));
  x=(double*)malloc(n*sizeof(double));
  y=(double*)malloc(n*sizeof(double));
  z=(double*)malloc(n*sizeof(double));

  /* Positions */

  fread(x,sizeof(double),n,fpi);
  fread(y,sizeof(double),n,fpi);
  fread(z,sizeof(double),n,fpi);

  /* Read accumulated averages/clear accumulators */

  if(nt>0) {

    fread(&accrc,sizeof(double),1,fpi);
    fread(&accrd,sizeof(double),1,fpi);
    fread(&accrm,sizeof(double),1,fpi);
    fread(&an,sizeof(double),1,fpi);
    fread(&an2,sizeof(double),1,fpi);
    fread(&au,sizeof(double),1,fpi);
    fread(&aw,sizeof(double),1,fpi);

    fread(ag,sizeof(long long),ndr,fpi);

  } else {

    accrc=0.0;
    accrd=0.0;
    accrm=0.0;
    an=0.0;
    an2=0.0;
    au=0.0;
    aw=0.0;

    for(i=0;i<=ndr-1;i++) {
      ag[i]=0;
    }

  }

  fclose(fpi);

  /* Parameters */

  nmax=n;

  c=cbrt(v);
  c2=0.5*c;
  c2m1=1.0/c2;
  drm1=1.0/dr;
  r2max=c2*c2;
  zv=zz*v;

  /* Cutoff & tail corrections */

  rc=0.5*c;
  rc2=rc*rc;

  su0=2.0*M_PI/v*(4.0/(9.0*pow(rc,9))-4.0/(3.0*pow(rc,3)));
  sw0=2.0*M_PI/v*(24.0/(3.0*pow(rc,3))-48.0/(9.0*pow(rc,9)));

  /* Initialize RNG */

  seed48(seed);

  /* Clear acceptance counters */

  naccc=0;
  naccd=0;
  naccm=0;

  return ;

}

/**********************************************************************/

void create() {

/**********************************************************************/

  int accept;
  int i,j;
  double dx,dy,dz,r2,rm6,su,xn,yn,zn;
  double *xt,*yt,*zt;

  /* Random position */

  xn=(drand48()-0.5)*c;
  yn=(drand48()-0.5)*c;
  zn=(drand48()-0.5)*c;

  /* Energy difference */

  su=(2*n+1)*su0;

  for(j=0;j<=n-1;j++) {
    dx=x[j]-xn;
    dx=dx-(int)(dx*c2m1)*c;
    dy=y[j]-yn;
    dy=dy-(int)(dy*c2m1)*c;
    dz=z[j]-zn;
    dz=dz-(int)(dz*c2m1)*c;
    r2=dx*dx+dy*dy+dz*dz;
    if(r2<rc2) {
      rm6=1.0/(r2*r2*r2);
      su=su+(4.0*rm6-4.0)*rm6;
    }
  }

  /* Acceptance test */

  if(su/t>alnmax) {
    accept=FALSE;
  } else {
    if(zv*exp(-su/t)/(n+1)>=1.0) {
      accept=TRUE;
    } else {
      if(drand48()<=zv*exp(-su/t)/(n+1)) {
	accept=TRUE;
      } else {
	accept=FALSE;
      }
    }
  }

  /* Accept move & append particle to arrays */

  if(accept) {

    naccc=naccc+1;

    /* Extend arrays */

    if(n>=nmax) {
      xt=(double*)malloc(n*sizeof(double));
      yt=(double*)malloc(n*sizeof(double));
      zt=(double*)malloc(n*sizeof(double));

      for(i=0;i<=n-1;i++) {
	xt[i]=x[i];
	yt[i]=y[i];
	zt[i]=z[i];
      }
      free(x);
      free(y);
      free(z);

      nmax=n+1;
      x=(double*)malloc(nmax*sizeof(double));
      y=(double*)malloc(nmax*sizeof(double));
      z=(double*)malloc(nmax*sizeof(double));

      for(i=0;i<=n-1;i++) {
	x[i]=xt[i];
	y[i]=yt[i];
	z[i]=zt[i];
      }
      free(xt);
      free(yt);
      free(zt);

    }

    x[n]=xn;
    y[n]=yn;
    z[n]=zn;
    n=n+1;

  }

  return;

}

/**********************************************************************/

void delete() {

/**********************************************************************/

  int accept;
  int i,j;
  double dx,dy,dz,r2,rm6,su;

  /* Empty box? */

  if(n<1) {
    return;
  }

  /* Random particle */

  i=min((int)(drand48()*n),n-1);

  /* Energy difference (ok for n=1) */

  su=-(2*n-1)*su0;

  for(j=0;j<=n-1;j++) {
    if(j!=i) {
      dx=x[j]-x[i];
      dx=dx-(int)(dx*c2m1)*c;
      dy=y[j]-y[i];
      dy=dy-(int)(dy*c2m1)*c;
      dz=z[j]-z[i];
      dz=dz-(int)(dz*c2m1)*c;
      r2=dx*dx+dy*dy+dz*dz;
      if(r2<rc2) {
	rm6=1.0/(r2*r2*r2);
	su=su-(4.0*rm6-4.0)*rm6;
      }
    }
  }

  /* Acceptance test */

  if(su/t>alnmax) {
    accept=FALSE;
  } else {
    if(n*exp(-su/t)/zv>=1.0) {
      accept=TRUE;
    } else {
      if(drand48()<=n*exp(-su/t)/zv) {
	accept=TRUE;
      } else {
	accept=FALSE;
      }
    }
  }

  /* Accept move & eliminate particle from arrays */

  if(accept) {
    naccd=naccd+1;
    x[i]=x[n-1];
    y[i]=y[n-1];
    z[i]=z[n-1];
    n=n-1;
  }

  return;

}

/**********************************************************************/

void move() {

/**********************************************************************/

  int accept;
  int i,j;
  double dx,dy,dz,r2,rm6,su,xn,yn,zn;

  /* Select random particle */

  i=min((int)(drand48()*n),n-1);

  /* Trial move (displacement) */

  xn=x[i]+(drand48()-0.5)*disp;
  xn=xn-(int)(xn*c2m1)*c;
  yn=y[i]+(drand48()-0.5)*disp;
  yn=yn-(int)(yn*c2m1)*c;
  zn=z[i]+(drand48()-0.5)*disp;
  zn=zn-(int)(zn*c2m1)*c;

  /* Energy difference */

  su=0.0;

  for(j=0;j<=n-1;j++) {
    if(j!=i) {

      /* New position */

      dx=x[j]-xn;
      dx=dx-(int)(dx*c2m1)*c;
      dy=y[j]-yn;
      dy=dy-(int)(dy*c2m1)*c;
      dz=z[j]-zn;
      dz=dz-(int)(dz*c2m1)*c;
      r2=dx*dx+dy*dy+dz*dz;
      if(r2<rc2) {
	rm6=1.0/(r2*r2*r2);
	su=su+(4.0*rm6-4.0)*rm6;
      }

      /* Old position */

      dx=x[j]-x[i];
      dx=dx-(int)(dx*c2m1)*c;
      dy=y[j]-y[i];
      dy=dy-(int)(dy*c2m1)*c;
      dz=z[j]-z[i];
      dz=dz-(int)(dz*c2m1)*c;
      r2=dx*dx+dy*dy+dz*dz;
      if(r2<rc2) {
	rm6=1.0/(r2*r2*r2);
	su=su-(4.0*rm6-4.0)*rm6;
      }

    }
  }

  /* Acceptance test */

  if(su<=0.0) {
    accept=TRUE;
  } else {
    if(su/t>alnmax) {
      accept=FALSE;
    } else {
      if(drand48()<=exp(-su/t)) {
	accept=TRUE;
      } else {
	accept=FALSE;
      }
    }
  }

  /* Accept move */

  if(accept) {
    naccm=naccm+1;
    x[i]=xn;
    y[i]=yn;
    z[i]=zn;
  }

  return;

}

/**********************************************************************/

void means() {

/**********************************************************************/

  int i,ir,j;
  double dx,dy,dz,r2,rm6,su,sw,p;

  /* Potential energy, virial & g(r) */

  su=n*n*su0;
  sw=n*n*sw0;

  for(i=0;i<=n-2;i++) {
    for(j=i+1;j<=n-1;j++) {

      dx=x[j]-x[i];
      dx=dx-(int)(dx*c2m1)*c;
      dy=y[j]-y[i];
      dy=dy-(int)(dy*c2m1)*c;
      dz=z[j]-z[i];
      dz=dz-(int)(dz*c2m1)*c;
      r2=dx*dx+dy*dy+dz*dz;

      if(r2<rc2) {
	rm6=1.0/(r2*r2*r2);
	su=su+(4.0*rm6-4.0)*rm6;
	sw=sw+(24.0-48.0*rm6)*rm6;
      }

      if(r2<r2max) {
	ir=(int)(drm1*sqrt(r2));
	if(ir<ndr) {
	  ag[ir]=ag[ir]+1;
	}
      }

    }
  }

  /* Print control variables */

  if((ntprint>0)&&(nt%ntprint==0)) {
    p=n/v*(t-sw/3.0);
    printf("%d %lf %lf %lf %d %lg %lg %lg\n",nt,\
	   naccc/max(pcre*n*ntskip,1.0),\
	   naccd/max(pcre*n*ntskip,1.0),\
	   naccm/max((1.0-2.0*pcre)*n*ntskip,1.0),\
	   n,n/v,su/n,p);
  }

  /* Accumulate averages */

  accrc=accrc+naccc/max(pcre*n*ntskip,1.0);
  accrd=accrd+naccd/max(pcre*n*ntskip,1.0);
  accrm=accrm+naccm/max((1.0-2.0*pcre)*n*ntskip,1.0);
  an=an+n;
  an2=an2+n*n;
  au=au+su;
  aw=aw+sw;

  /* Clear acceptance counters */

  naccc=0;
  naccd=0;
  naccm=0;

  return;

}

/**********************************************************************/

void putcf() {

/**********************************************************************/

  FILE *fpo;
  unsigned short dummy[3];
  unsigned short *seed;

  /* RNG seed */

  seed=seed48(dummy);

  /* Write checkpoint file */

  fpo=fopen("gcmclj_out.dat","w");

  fwrite(&n,sizeof(int),1,fpo);
  fwrite(&nt,sizeof(int),1,fpo);
  fwrite(&ntjob,sizeof(int),1,fpo);
  fwrite(&ntprint,sizeof(int),1,fpo);
  fwrite(&ntskip,sizeof(int),1,fpo);
  fwrite(seed,sizeof(unsigned short),3,fpo);

  fwrite(&disp,sizeof(double),1,fpo);
  fwrite(&dr,sizeof(double),1,fpo);
  fwrite(&pcre,sizeof(double),1,fpo);
  fwrite(&t,sizeof(double),1,fpo);
  fwrite(&v,sizeof(double),1,fpo);
  fwrite(&zz,sizeof(double),1,fpo);

  fwrite(x,sizeof(double),n,fpo);
  fwrite(y,sizeof(double),n,fpo);
  fwrite(z,sizeof(double),n,fpo);

  fwrite(&accrc,sizeof(double),1,fpo);
  fwrite(&accrd,sizeof(double),1,fpo);
  fwrite(&accrm,sizeof(double),1,fpo);
  fwrite(&an,sizeof(double),1,fpo);
  fwrite(&an2,sizeof(double),1,fpo);
  fwrite(&au,sizeof(double),1,fpo);
  fwrite(&aw,sizeof(double),1,fpo);

  fwrite(ag,sizeof(long long),ndr,fpo);

  fclose(fpo);

  /* Deallocate arrays */

  free(ag);
  free(x);
  free(y);
  free(z);

  return;

}

/**********************************************************************/

int main() {

/**********************************************************************/

  int i,j;
  double ran;

  /* Read startup/checkpoint file & initialize variables */

  getcf();

  /* Do ntskip*ntjob passes (particle creations/deletions/moves) */

  for(i=1;i<=ntjob;i++) {
    nt=nt+1;
    for(j=1;j<=n*ntskip;j++) {
      ran=drand48();
      if(ran<pcre) {
	create();               /* Create particle */
      } else if(ran<2.0*pcre) {
	delete();               /* Delete particle */
      } else {
	move();                 /* Move particle */
      }
    }
    means();
  }

  /* Write checkpoint file */

  putcf();

  return 0;

}

/**********************************************************************/
